src/library/autodj/autodjfeature.cpp

```diff
@@ -275,41 +271,57 @@ void AutoDJFeature::slotCrateAutoDjChanged(int crateId, bool added) {
             }
         }
     }
-    // Let subscribers know whether it's possible to add a random track.
-    emit(enableAddRandom(m_crateList.length() > 0));
 #endif // __AUTODJCRATES__
 }
+// Adds a random track : this will be faster when there are sufficiently large 
+// tracks in the crates
 
 void AutoDJFeature::slotAddRandomTrack(bool) {
 #ifdef __AUTODJCRATES__
-    int retriveAttempts = 0;
-    while (retriveAttempts < kMaxRetiveAttempts) {
-        // Get access to the auto-DJ playlist.
-        PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
-        if (m_iAutoDJPlaylistId >= 0) {
+    int failedRetriveAttempts = 0;
+    // Get access to the auto-DJ playlist
+    PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
+    int iTrackId = -1;
+
+    if (m_iAutoDJPlaylistId >= 0) {
+        while (failedRetriveAttempts < kMaxRetriveAttempts) {
             // Get the ID of a randomly-selected track.
-            int iTrackId = m_autoDjCratesDao.getRandomTrackId();
-            //qDebug() << "we got iTrackid " <<iTrackId;
+            iTrackId = m_autoDjCratesDao.getRandomTrackId();
             if (iTrackId != -1) {
                 // Get Track Information
                 TrackPointer addedTrack = (m_pTrackCollection->getTrackDAO()).getTrack(iTrackId);
                 if(addedTrack->exists()) {
-                    //qDebug() << addedTrack->getTitle() << " Exists";
-                    // Add this randomly-selected track to the auto-DJ playlist.
                     playlistDao.appendTrackToPlaylist(iTrackId, m_iAutoDJPlaylistId);
                     m_pAutoDJView->onShow();
                     return;
                 }
-                else {
-                    qDebug () <<"Track does not exist " << addedTrack->getTitle();
+            }
+            failedRetriveAttempts += 1;
+        }
+        // If we couldn't get a track from the crates , get one from the library
+        if (failedRetriveAttempts == kMaxRetriveAttempts || iTrackId == -1) {
```

Comment:
hm actually this whole if statement is unneeded since the function will return if it finds a valid track.  So just get rid of it